### PR TITLE
add simple validation cache

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
When creating more than 3 machines at once, we run into rate-limits by aws.
As it does not make sense to validate a machine which has already been validated, this introduces a simple key-value cache for validation.
The machine-UID+machine-Revision will be used for identification. A key will get invalidated if the further processing fails.

### CODE REVIEW

File: pkg/controller/machine.go [hunk 1]
```diff
@@ -426,16 +429,27 @@ func (c *Controller) createInstanceForMachine(prov cloud.Provider, machine *mach
 		}
 		glog.V(4).Infof("Successfully updated machine '%s' with defaults!", machine.Name)
 	}
-	if err := c.validateMachine(prov, machine); err != nil {
-		if _, errNested := c.updateMachineError(machine, machinev1alpha1.InvalidConfigurationMachineError, err.Error()); errNested != nil {
-			return fmt.Errorf("failed to update machine error after failed validation: %v", errNested)
+
+	cacheKey := string(machine.UID) + machine.ResourceVersion
+	if !c.validationCache[cacheKey] {
```
Comment (p0lyn0mial) [inline pkg/controller/machine.go]: I think that we also want to check the value stored under `cacheKey`

File: pkg/controller/machine.go [hunk 2]
```diff
@@ -426,16 +429,27 @@ func (c *Controller) createInstanceForMachine(prov cloud.Provider, machine *mach
...
+		c.validationCache[cacheKey] = true
+	} else {
+		glog.V(6).Infof("Skipping validation as the machine was already successfully validated before")
 	}
 	providerInstance, err := prov.Get(machine)
 
 	// case 2: retrieving instance from provider was not successful
 	if err != nil {
+		//First invalidate the validation cache to make sure we run the validation on the next sync.
+		//This might happen in case the user invalidates his provider credentials...
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/controller/machine.go]: do you think we could use build-in `delete` function to remove the keys from the cache ?

[the remark this answers is not in the document]
Comment (mrIncompetent) [reply pkg/controller/machine.go]: For cleanup purpose?
Why not simply overriding the map with a new one?

File: pkg/controller/machine.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/controller/machine.go]: I just realised that this line will check both: the key and the value

**Pull request merged**